native_client/deepspeech.cc

```diff
@@ -8,45 +8,423 @@
+   * @param logits         Matrix of logits, of dimensions:
+   *                       [n_frames][batch_size][num_classes]
+   *
+   * @return String representing the decoded text.
+   *
+   * @note This method will free @p logits.
+   */
+  char* decode(std::vector<float>& logits);
+
+  /**
+   * @brief Do a single inference step in the acoustic model, with:
+   *          input=mfcc
+   *          input_lengths=[n_frames]
+   *
+   * @param mfcc batch input data
+   * @param n_frames number of timesteps in the data
+   *
+   * @param[out] output_logits Should be large enough to fit
+   *                           aNFrames * alphabet_size floats.
+   */
+  void infer(float* mfcc, int n_frames, std::vector<float>& output_logits);
+
+  void processAudioWindow(StreamingState* ctx, short* buf, unsigned int len);
+  void processMfccWindow(StreamingState* ctx, float* buf, unsigned int len);
+  void addZeroMfccWindow(StreamingState* ctx);
+  void processBatch(StreamingState* ctx, float* buf, unsigned int len);
 };
 
+StreamingState*
+Private::setupStream(unsigned int prealloc_frames,
+                     unsigned int /*sample_rate*/)
+{
+  Status status = session->Run({}, {}, {"initialize_state"}, nullptr);
+
+  if (!status.ok()) {
+    std::cerr << "Error running session: " << status << "\n";
+    return nullptr;
+  }
+
+  StreamingState* ctx = new StreamingState;
+  const size_t num_classes = alphabet->GetSize() + 1; // +1 for blank
+
+  ctx->accumulated_logits.reserve(prealloc_frames * BATCH_SIZE * num_classes);
+
+  memset(ctx->audio_buffer, 0, sizeof(ctx->audio_buffer));
+  ctx->audio_buffer_len = 0;
+
+  memset(ctx->mfcc_buffer, 0, sizeof(ctx->mfcc_buffer));
+  ctx->mfcc_buffer_len = MFCC_FEATURES*MFCC_CONTEXT; // initial zero entries
+
+  memset(ctx->batch_buffer, 0, sizeof(ctx->batch_buffer));
+  ctx->batch_buffer_len = 0;
+
+  ctx->skip_next_mfcc = false;
+
+  return ctx;
+}
+
+void
+Private::feedAudioContent(StreamingState* ctx,
+                          const short* buffer,
+                          unsigned int buffer_size)
+{
+  while (buffer_size > 0) {
+    while (buffer_size > 0 && ctx->audio_buffer_len < AUDIO_WIN_LEN_SAMPLES) {
+      ctx->audio_buffer[ctx->audio_buffer_len] = *buffer;
+      ++ctx->audio_buffer_len;
+      ++buffer;
+      --buffer_size;
+    }
+
+    if (ctx->audio_buffer_len == AUDIO_WIN_LEN_SAMPLES) {
+      processAudioWindow(ctx, ctx->audio_buffer, ctx->audio_buffer_len);
+      // shift data by one step of 10ms
+      memmove(ctx->audio_buffer, ctx->audio_buffer + AUDIO_WIN_STEP_SAMPLES, (AUDIO_WIN_LEN_SAMPLES - AUDIO_WIN_STEP_SAMPLES) * sizeof(short));
+      ctx->audio_buffer_len -= AUDIO_WIN_STEP_SAMPLES;
+    }
+  }
+}
+
+char*
+Private::finishStream(StreamingState* ctx)
+{
+  // flush audio buffer
+  processAudioWindow(ctx, ctx->audio_buffer, ctx->audio_buffer_len);
+
+  // add empty mfcc vectors at end of sample
+  for (int i = 0; i < MFCC_CONTEXT; ++i) {
+    addZeroMfccWindow(ctx);
+  }
```

Comment:
Why do we add with those empty MFCC ? I see we put one full context, I would have expected we perform some padding but not a full blank context?

Reply:
That's just matching the processing we do for training. The context window is 9 frames in the past and 9 frames in the future, so we need 9 zero windows in the beginning (done in `Private::setupStream`) and 9 zero windows in the end, in order to capture the first and last frames of the audio.